include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,1386 @@
+			rem = vec3( irr_glsl_ggx_aniso_dielectric_cos_remainder_and_pdf(pdf, s, currInteraction, uf, ior[0].x, ax, ay) );
+		}
+		else 
+#endif
+		{
+			s = irr_glsl_ggx_cos_generate(currInteraction, u.xy, ax, ay, uf);
+			rem = irr_glsl_ggx_aniso_cos_remainder_and_pdf(pdf, s, currInteraction, uf, ior, ax, ay);
+		}
+	} else
+#endif //NDF_GGX
+
+#ifdef NDF_BECKMANN
+	if (ndf == NDF_BECKMANN) {
+#if defined(OP_THINDIELECTRIC) || defined(OP_DIELECTRIC)
+		if (is_bsdf) {
+			s = irr_glsl_beckmann_dielectric_cos_generate(currInteraction, u, ax, ay, ior[0].x, uf);
+			rem = vec3(irr_glsl_beckmann_aniso_dielectric_cos_remainder_and_pdf(pdf, s, currInteraction, uf, ior[0].x, ax, ay) );
+		}
+		else
+#endif
+		{
+			s = irr_glsl_beckmann_cos_generate(currInteraction, u.xy, ax, ay, uf);
+			rem = irr_glsl_beckmann_aniso_cos_remainder_and_pdf(pdf, s, currInteraction, uf, ior, ax, ay);
+		}
+	} else
+#endif //NDF_BECKMANN
+
+#ifdef NDF_PHONG
+	if (ndf == NDF_PHONG) {
+#if defined(OP_THINDIELECTRIC) || defined(OP_DIELECTRIC)
+		if (is_bsdf) {
+			s = irr_glsl_beckmann_dielectric_cos_generate(currInteraction, u, ax, ay, ior[0].x, uf);
+			rem = vec3( irr_glsl_beckmann_aniso_dielectric_cos_remainder_and_pdf(pdf, s, currInteraction, uf, ior[0].x, ax, ay) );
+		}
+		else
+#endif
+		{
+			s = irr_glsl_beckmann_cos_generate(currInteraction, u.xy, ax, ay, uf);
+			rem = irr_glsl_beckmann_aniso_cos_remainder_and_pdf(pdf, s, currInteraction, uf, ior, ax, ay);
+		}
+	} else
+#endif //NDF_PHONG
+	{}
+
+#ifdef OP_PLASTIC
+	if (is_plastic)
+	{
+		vec3 eval;
+		float pdf_b;
+		float wa;
+		float wb;
+		if (u.z>=plastic_spec_weight) { //chose diffuse as generator
+			wb = plastic_spec_weight;
+			wa = 1.0-wb;
+
+			BEGIN_CASES(ndf)
+#ifdef NDF_GGX
+			CASE_BEGIN(ndf, NDF_GGX) {
+				eval = irr_glsl_ggx_height_correlated_aniso_cos_eval(s, currInteraction, uf, ior, ax, ay);
+				pdf_b = irr_glsl_ggx_pdf(currInteraction, uf, ax, ay, ax2, ay2);
+			} CASE_END
+#endif //NDF_GGX
+
+#ifdef NDF_BECKMANN
+			CASE_BEGIN(ndf, NDF_BECKMANN) {
+				eval = irr_glsl_beckmann_aniso_height_correlated_cos_eval(s, currInteraction, uf, ior, ax, ay);
+				pdf_b = irr_glsl_beckmann_pdf(currInteraction, uf, ax, ax2, ay, ay2);
+			} CASE_END
+#endif //NDF_BECKMANN
+
+#ifdef NDF_PHONG
+			CASE_BEGIN(ndf, NDF_PHONG) {
+				eval = irr_glsl_beckmann_aniso_height_correlated_cos_eval(s, currInteraction, uf, ior, ax, ay);
+				pdf_b = irr_glsl_beckmann_pdf(currInteraction, uf, ax, ax2, ay, ay2);
+			} CASE_END
+#endif //NDF_PHONG
+			CASE_OTHERWISE
+			{}
+			END_CASES
+		}
+		else { //specular was generator
+			wa = plastic_spec_weight;
+			wb = 1.0-wa;
+			eval = refl*irr_glsl_oren_nayar_cos_eval(s, currInteraction.isotropic, ax2);//TODO fresnels and correction
+			pdf_b = irr_glsl_oren_nayar_pdf(s, currInteraction.isotropic);
+		}
+
+		rem = (rem*wa + eval/pdf*wb)/(wa + pdf_b/pdf*wb);
+		pdf = pdf*wa + pdf_b*wb;
+	}
+#endif
+
+	out_remainder = weight*rem;
+	out_pdf = weight*pdf;
```

Comment:
this is wrong, the `remainder` shouldn't be multiplied by the weight.

`pdf` can be multiplied by `weight` or you could not track weight at all and use `rcpChoiceProb` directly

Reply:
`weight` is a value resulting from traversing possible chain of blends (choosing left/right). Im multiplying both `rem` and `pdf` by the `weight` so that in the end of `runGenerateAndRemainderStream()` i have `generator_rem` and `generator_pdf` properly weighted. I dont really understand why generator remainder shouldnt be weighted if OP_BLEND is essentially `mix(vec4(rem1, pdf1), vec4(rem2, pdf2))`

```
vec3 runGenerateAndRemainderStream(in MC_precomputed_t precomp, in instr_stream_t gcs, in instr_stream_t rnps, in vec3 rand, out float out_pdf, out irr_glsl_LightSample out_smpl)
{
	instr_t generator;
	vec3 generator_rem;
	float generator_pdf;
	irr_glsl_AnisotropicMicrofacetCache microfacet;
	irr_glsl_LightSample s = irr_bsdf_cos_generate(precomp, gcs, rand, generator_rem, generator_pdf, generator, microfacet);
	eval_and_pdf_t eval_pdf = irr_bsdf_eval_and_pdf(precomp, rnps, s, generator, microfacet);
	bxdf_eval_t acc = eval_pdf.rgb;
	float restPdf = eval_pdf.a;
	float pdf = generator_pdf + restPdf;

	out_smpl = s;
	out_pdf = pdf;

	vec3 rem = generator_rem/(1.0 + restPdf/generator_pdf) + acc/pdf;

	return rem;
}
```

Reply:
Because the estimator (overall remainder) is
`rem = Sum{0...N}(w_i*f_i)/Sum{0...N}(w_i*p_i);`

you can split it w.r.t. the generator `g`
`rem = (w_g*f_g+Sum{i!=g}(w_i*f_i))/Sum{0...N}(w_i*p_i);`

divide both sides by `w_g*p_g`
`rem = (w_g*f_g/(w_g*p_g)+Sum{i!=g}(w_i*f_i)/(w_g*p_g))/(Sum{i!=g}(w_i*p_i)/(w_g*p_g)+1.0);`

and it just so happens that
`w_g*f_g/(w_g*p_g)`
is the unweighted remainder of the generator BxDF
`f_g/p_g`

Reply:
btw there's two forms of getting the remainder
`generator_rem/(1.0+restPdf/generator_pdf) + acc/pdf`
and
`(generator_rem+acc/generator_pdf)/(1.0+restPdf/generator_pdf)`

where `generator_rem` is the unweighted remainder of the generator